provider/v2/azure/discoverer/discoverer.go

```diff
@@ -17,15 +17,142 @@ package discoverer
+	"github.com/openclarity/vmclarity/core/log"
+	"github.com/openclarity/vmclarity/core/to"
 	"github.com/openclarity/vmclarity/provider"
 )
 
-var _ provider.Discoverer = &Discoverer{}
-
-type Discoverer struct{}
+type Discoverer struct {
+	VMClient    *armcompute.VirtualMachinesClient
+	DisksClient *armcompute.DisksClient
+}
 
 func (d *Discoverer) DiscoverAssets(ctx context.Context) provider.AssetDiscoverer {
-	// TODO implement me
-	panic("implement me")
+	assetDiscoverer := provider.NewSimpleAssetDiscoverer()
+
+	go func() {
+		defer close(assetDiscoverer.OutputChan)
+
+		// list all vms in all resourceGroups in the subscription
+		res := d.VMClient.NewListAllPager(nil)
+		for res.More() {
+			page, err := res.NextPage(ctx)
+			if err != nil {
+				assetDiscoverer.Error = fmt.Errorf("failed to get next page: %w", err)
+				return
+			}
+			ts, err := d.processVirtualMachineListIntoAssetTypes(ctx, page.VirtualMachineListResult)
+			if err != nil {
+				assetDiscoverer.Error = err
+				return
+			}
+
+			for _, asset := range ts {
+				select {
+				case assetDiscoverer.OutputChan <- asset:
+				case <-ctx.Done():
+					assetDiscoverer.Error = ctx.Err()
+					return
+				}
+			}
+		}
+	}()
+
+	return assetDiscoverer
+}
+
+func (d *Discoverer) processVirtualMachineListIntoAssetTypes(ctx context.Context, vmList armcompute.VirtualMachineListResult) ([]apitypes.AssetType, error) {
+	ret := make([]apitypes.AssetType, 0, len(vmList.Value))
+	for _, vm := range vmList.Value {
+		info, err := getVMInfoFromVirtualMachine(vm, d.getRootVolumeInfo(ctx, vm))
+		if err != nil {
+			return nil, fmt.Errorf("unable to convert instance to vminfo: %w", err)
+		}
+		ret = append(ret, info)
+	}
+	return ret, nil
+}
+
+func (d *Discoverer) getRootVolumeInfo(ctx context.Context, vm *armcompute.VirtualMachine) *apitypes.RootVolume {
+	logger := log.GetLoggerFromContextOrDiscard(ctx)
+	ret := &apitypes.RootVolume{
+		SizeGB:    int(to.ValueOrZero(vm.Properties.StorageProfile.OSDisk.DiskSizeGB)),
+		Encrypted: apitypes.RootVolumeEncryptedUnknown,
+	}
+	osDiskID, err := arm.ParseResourceID(to.ValueOrZero(vm.Properties.StorageProfile.OSDisk.ManagedDisk.ID))
+	if err != nil {
+		logger.Warnf("Failed to parse disk ID. DiskID=%v: %v",
+			to.ValueOrZero(vm.Properties.StorageProfile.OSDisk.ManagedDisk.ID), err)
+		return ret
+	}
+	osDisk, err := d.DisksClient.Get(ctx, osDiskID.ResourceGroupName, osDiskID.Name, nil)
+	if err != nil {
+		logger.Warnf("Failed to get OS disk. DiskID=%v: %v",
+			to.ValueOrZero(vm.Properties.StorageProfile.OSDisk.ManagedDisk.ID), err)
+		return ret
+	}
+	ret.Encrypted = isEncrypted(osDisk)
+	ret.SizeGB = int(to.ValueOrZero(osDisk.Disk.Properties.DiskSizeGB))
+
+	return ret
+}
+
+func getVMInfoFromVirtualMachine(vm *armcompute.VirtualMachine, rootVol *apitypes.RootVolume) (apitypes.AssetType, error) {
+	assetType := apitypes.AssetType{}
+	err := assetType.FromVMInfo(apitypes.VMInfo{
+		ObjectType:       "VMInfo",
```

Comment:
note: we should probably make these items constants in `vmclarity/api/types` in another PR

Reply:
created https://github.com/openclarity/vmclarity/issues/1367